mayastor/src/bdev/dev/nvmx/controller.rs

```diff
@@ -0,0 +1,1335 @@
+    // As of now, only 1 namespace per controller is supported.
+    pub fn namespace(&self) -> Option<Arc<NvmeNamespace>> {
+        let inner = self
+            .inner
+            .as_ref()
+            .expect("(BUG) no inner NVMe controller defined yet");
+
+        if let Some(ns) = inner.namespaces.get(0) {
+            Some(Arc::clone(ns))
+        } else {
+            debug!("no namespaces associated with the current controller");
+            None
+        }
+    }
+
+    /// we should try to avoid this
+    pub fn ctrlr_as_ptr(&self) -> *mut spdk_nvme_ctrlr {
+        self.inner.as_ref().map_or(std::ptr::null_mut(), |c| {
+            let ptr = c.ctrlr.as_ptr();
+            debug!("SPDK handle {:p}", ptr);
+            ptr
+        })
+    }
+
+    /// Register callbacks.
+    fn register_callbacks(&mut self) {
+        let ctrlr = self.ctrlr_as_ptr();
+
+        unsafe {
+            spdk_nvme_ctrlr_register_aer_callback(
+                ctrlr,
+                Some(aer_cb),
+                ctrlr as *mut c_void,
+            );
+        };
+    }
+
+    /// populate name spaces, current we only populate the first namespace
+    fn populate_namespaces(&mut self) -> bool {
+        let ctrlr = self.ctrlr_as_ptr();
+        let mut ctrlr_inner = self.inner.as_mut().unwrap();
+        let ns = unsafe { spdk_nvme_ctrlr_get_ns(ctrlr, 1) };
+        let ns_active = unsafe { spdk_nvme_ctrlr_is_active_ns(ctrlr, 1) };
+        let mut notify_listeners = false;
+
+        // Deactivate existing namespace in case it is no longer active.
+        if !ns_active && ctrlr_inner.namespaces.len() != 0 {
+            info!("{}: deactivating existing namespace", self.name);
+            notify_listeners = true;
+        }
+
+        let namespaces = if ns.is_null() || !ns_active {
+            warn!(
+                "{}: no active namespaces reported by the NVMe controller",
+                self.name
+            );
+            vec![]
+        } else {
+            info!("{}: namespace successfully populated", self.name);
+            vec![Arc::new(NvmeNamespace::from_ptr(ns))]
+        };
+
+        ctrlr_inner.namespaces = namespaces;
+
+        // Fault the controller in case of inactive namespace.
+        if !ns_active {
+            self
+                .state_machine
+                .transition(Faulted(ControllerFailureReason::NamespaceInitFailed))
+                .expect("failed to fault controller in response to ns enumeration failure");
+        }
+
+        // Notify listeners in case of namespace removal.
+        if notify_listeners {
+            self.notify_event(DeviceEventType::DeviceRemoved);
+        }
+
+        ns_active
+    }
+
+    /// Get controller state.
+    pub fn get_state(&self) -> NvmeControllerState {
+        self.state_machine.current_state()
+    }
+
+    /// Reset the controller.
+    /// Upon reset all pending I/O operations are cancelled and all I/O handles
+    /// are reinitialized.
+    pub fn reset(
```

Comment:
Does pending IO's include IOs which have already been submitted? So unlike when using spdk the IOs won't automatically be retried and will instead fail?

Reply:
We don't know if an IO is submitted or not at this point, and it's not possible to find out either. The only thing we do know, and what is imported, is if the IO completed. If it is in this list -- it is not.